rpcs3/Emu/Io/RB3MidiGuitar.cpp

```diff
@@ -0,0 +1,245 @@
+	{
+		switch (buf[2])
+		{
+		case 0x89:
+			rb3_midi_guitar_log.notice("MIDI data enabled.");
+			buttons_enabled = true;
+			response_pos = 0;
+			break;
+		case 0x81:
+			rb3_midi_guitar_log.notice("MIDI data disabled.");
+			buttons_enabled = false;
+			response_pos = 0;
+			break;
+		default:
+			rb3_midi_guitar_log.warning("Unhandled SET_REPORT request: 0x%02X");
+			break;
+		}
+	}
+	// the game expects some sort of response to the configutarion packet
+	else if (bmRequestType == 0xa1 && bRequest == 0x1)
+	{
+		transfer->expected_count = buf_size;
+		// TODO: ensure memory safety
+		if (buttons_enabled)
+		{
+			memcpy(buf, &enabled_response[response_pos], buf_size);
+			response_pos += buf_size;
+		}
+		else
+		{
+			memcpy(buf, &disabled_response[response_pos], buf_size);
+			response_pos += buf_size;
+		}
+	}
+	else if (bmRequestType == 0x21 && bRequest == 0x9 && wLength == 8)
+	{
+		// the game uses this request to do things like set the LEDs
+		// we don't have any LEDs, so do nothing
+	}
+	else
+	{
+		usb_device_emulated::control_transfer(bmRequestType, bRequest, wValue, wIndex, wLength, buf_size, buf, transfer);
+	}
+}
+
+void usb_device_rb3_midi_guitar::interrupt_transfer(u32 buf_size, u8* buf, u32 /*endpoint*/, UsbTransfer* transfer)
+{
+	transfer->fake = true;
+	transfer->expected_count = buf_size;
+	transfer->expected_result = HC_CC_NOERR;
+	// the real device takes 8ms to send a response, but there is
+	// no reason we can't make it faster
+	transfer->expected_time = get_timestamp() + 1'000;
+
+	// default input state
+	u8 bytes[27] = {
+		0x00, 0x00, 0x08, 0x80, 0x80, 0x00, 0x00, 0x00,
+		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x40,
+		0x40, 0x40, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+		0x00, 0x00, 0x00};
+
+	memcpy(buf, bytes, 27);
+
+	while (true)
+	{
+		u8 midi_msg[32];
+		usz size = sizeof(midi_msg);
+
+		// this returns a double as some sort of delta time, with -1.0
+		// being used to signal an error
+		if (rtmidi_in_get_message(midi_in, midi_msg, &size) == -1.0)
+		{
+			rb3_midi_guitar_log.error("Error getting midi message");
+			return;
+		}
+
+		if (size == 0)
+		{
+			break;
+		}
+
+		parse_midi_message(midi_msg, size);
+	}
+
+	write_state(buf);
+}
+
+void usb_device_rb3_midi_guitar::parse_midi_message(u8 msg[32], usz size)
+{
+	// this is not emulated correctly but the game doesn't seem to care
+	button_state.count++;
+
+	rb3_midi_guitar_log.success("msg: %d", msg[0]);
+	// read frets
+	if (size == 8 && msg[0] == 0xF0 && msg[4] == 0x01)
+	{
+		switch (msg[5])
+		{
+		case 1:
+			button_state.frets[0] = msg[6] - 0x40;
+			break;
+		case 2:
+			button_state.frets[1] = msg[6] - 0x3B;
+			break;
+		case 3:
+			button_state.frets[2] = msg[6] - 0x37;
+			break;
+		case 4:
+			button_state.frets[3] = msg[6] - 0x32;
+			break;
+		case 5:
+			button_state.frets[4] = msg[6] - 0x2D;
+			break;
+		case 6:
+			button_state.frets[5] = msg[6] - 0x28;
+			break;
+		}
```

Comment:
Mising default

Reply:
That's irrelevant.
All switches that don't handle all their possible cases need a default, otherwise you get compiler warnings.